synapse/storage/data_stores/main/push_rule.py

```diff
@@ -656,12 +662,31 @@ def set_push_rule_enabled(self, user_id, rule_id, enabled):
                 user_id,
                 rule_id,
                 enabled,
+                is_default_rule,
             )
 
     def _set_push_rule_enabled_txn(
-        self, txn, stream_id, event_stream_ordering, user_id, rule_id, enabled
+        self,
+        txn,
+        stream_id,
+        event_stream_ordering,
+        user_id,
+        rule_id,
+        enabled,
+        is_default_rule,
     ):
         new_id = self._push_rules_enable_id_gen.get_next()
+
+        if not is_default_rule:
+            try:
+                # first check it exists
+                self.db.simple_select_one_onecol_txn(
```

Comment:
I'm not sure about this, but you might need to `SELECT FOR UPDATE` to ensure that, if an update races with a delete, we don't end up with a dangling `enabled` row.

possibly best to discuss in #synapse-dev

Reply:
bump